### PR TITLE
[feature] support dpo trainer

### PR DESCRIPTION
**Pull request by yezhem**

### CODE REVIEW

**Comment by yezhem** [comment]
@waitfor-night can you check this code about the dpo algorithm?

**Review by waitfor-night — CHANGES_REQUESTED** [review]
Does it support importing a trained LoRA weight as the  reference model and initialization for the policy model?

**Comment by yezhem** [comment]
@waitfor-night I added the feature to load the trained LoRA adapter and freeze it when trained by DPO can you check and test it?

File: mlora/executor/context/lora.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            if config.target_[target_name] is not True:
+                continue
+
+            self.adapter_model_[linear_name] = LoRA(config.name_,
+                                                    linear_info.in_dim_, linear_info.out_dim_,
+                                                    config.r_, config.alpha_, config.dropout_)
+
+        self.__init_weight()
+
+        self.device_ = "cpu"
```
Comment (merlintang) [inline mlora/executor/context/lora.py:34]: default is the cpu?

File: mlora/executor/task/dpo_task.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        # include reference and policy models' chosen and reject data
+        assert len(ret_tokens) % 4 == 0
+
+        ref_start_idx = start_idx
+        ref_end_idx = ref_start_idx + len(ref_model_token)
+
+        policy_start_idx = ref_end_idx
+        policy_end_idx = policy_start_idx + len(policy_model_token)
+
+        def loss_fn(input: torch.Tensor, target: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
```
Comment (merlintang) [inline mlora/executor/task/dpo_task.py:109]: can you provide the related forumal for this loss function

File: mlora/prompter/dpo_prompter.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+from .prompter import Prompter
+
+
+from typing import Dict, List, Tuple
+
+
+class DPOPrompter(Prompter):
```
Comment (merlintang) [inline mlora/prompter/dpo_prompter.py:7]: can you give the description for this clas?

**Review by merlintang — APPROVED** [review]

**Comment by merlintang** [comment]
leave comments

File: tests/dpo/dpo_case_2.yaml
Comment (waitfor-night) [inline tests/dpo/dpo_case_2.yaml:1]: When using a pre-trained adapter as the initialization for the policy model, the process can be somewhat cumbersome, but it is not logically difficult. It is recommended to modify the YAML file in this case.

**Review by waitfor-night — APPROVED** [review]

File: mlora/executor/context/lora.py (hunk 1 above)
Comment (yezhem) [reply mlora/executor/context/lora.py:34]: yes, only scheduling the task in the running queue, will change the weight to the GPU for train.

File: mlora/executor/task/dpo_task.py (hunk 2 above)
Comment (yezhem) [reply mlora/executor/task/dpo_task.py:109]: yes, we will update in document.

File: mlora/executor/task/dpo_task.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            mask = ~mask[ref_start_idx:policy_end_idx]
+
+            logits = input[ref_start_idx:policy_end_idx][:,
+                                                         :-1, :].log_softmax(-1)
+            labels = target[ref_start_idx:policy_end_idx][:, 1:].to(
+                input.device)
+            mask = mask.long()[:, 1:].to(input.device)
+
+            pre_token_logps = torch.gather(
```
Comment (waitfor-night) [inline mlora/executor/task/dpo_task.py:118]: ”pre_token_logps“ should be "per_token_logps"
